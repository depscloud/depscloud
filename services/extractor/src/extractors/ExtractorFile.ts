import cheerio = require("cheerio");
import YAML = require("js-yaml");
import TOML = require("toml");

export default class ExtractorFile {
    private readonly body: string;

    constructor(body: string) {
        this.body = body;
    }

    public raw(): string {
        return this.body;
    }

    public json(): any {
        return JSON.parse(this.body);
    }

    public toml(): any {
        return TOML.parse(this.body);
    }

    public yaml(): any {
        return YAML.safeLoad(this.body);
    }

    public xml(): cheerio.Cheerio {
        return cheerio.load(this.body).root();
    }
}
