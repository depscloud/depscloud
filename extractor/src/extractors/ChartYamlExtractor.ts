import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import {ManifestDependency, ManifestFile} from "@depscloud/api/v1beta";
import MatchConfig from "../matcher/MatchConfig";
import Languages from "./Languages";

export default class ChartYamlExtractor implements Extractor {
    public async extract(url: string, files: { [p: string]: ExtractorFile }): Promise<ManifestFile> {
        const {
            name,
            version,
            dependencies: deps,
        } = files["Chart.yaml"].yaml();

        const dependencies: Array<ManifestDependency> = (deps || []).map((dep) => {
            let fullName = String(dep.repository || "").trim();
            if (!fullName.startsWith("file://")) {
                if (fullName[fullName.length - 1] !== "/") {
                    fullName += "/";
                }
                fullName += dep.name;
            }

            return {
                name: fullName,
                versionConstraint: dep.version,
                scopes: dep.condition ? [ dep.condition ] : [],
            }
        })

        return {
            language: Languages.HELM,
            system: Languages.HELM,
            name,
            version,
            sourceUrl: "",
            dependencies,
        }
    }

    public matchConfig(): MatchConfig {
        return {
            includes: [ "Chart.yaml" ],
            excludes: [],
        };
    }

    public requires(): string[] {
        return [ "Chart.yaml" ];
    }
}