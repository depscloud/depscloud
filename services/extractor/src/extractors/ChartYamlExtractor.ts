import Extractor from "./Extractor";
import ExtractorFile from "./ExtractorFile";
import {ManifestDependency, ManifestFile} from "@depscloud/api/v1beta";
import MatchConfig from "../matcher/MatchConfig";
import Languages from "./Languages";

export default class ChartYamlExtractor implements Extractor {
    public async extract(url: string, files: { [p: string]: ExtractorFile }): Promise<ManifestFile> {
        const {
            name,
            version,
            dependencies: deps,
        } = files["Chart.yaml"].yaml();

        const dependencies: Array<ManifestDependency> = (deps || []).map((dep) => {
            let versionConstraint = dep.repository || "";
            if (versionConstraint !== "" && dep.version) {
                versionConstraint += "#" + dep.version
            }

            return {
                name: dep.name,
                versionConstraint,
                scopes: dep.condition ? [ dep.condition ] : [],
            }
        })

        return {
            language: Languages.HELM,
            system: Languages.HELM,
            name,
            version,
            sourceUrl: "",
            dependencies,
        }
    }

    public matchConfig(): MatchConfig {
        return {
            includes: [ "**/Chart.yaml" ],
            excludes: [],
        };
    }

    public requires(): string[] {
        return [ "Chart.yaml" ];
    }
}