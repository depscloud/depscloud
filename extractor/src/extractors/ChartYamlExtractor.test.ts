import {readFile} from "fs";
import {promisify} from "util";
import ChartYamlExtractor from "./ChartYamlExtractor";
import ExtractorFile from "./ExtractorFile";

const readFileAsync = promisify(readFile);

describe("ChartYamlExtractor", () => {
    test("fullParse", async () => {
        const filePath = require.resolve("./testdata/Chart.yaml");
        const buffer = await readFileAsync(filePath);
        const content = buffer.toString();

        const parser = new ChartYamlExtractor();

        const actual = await parser.extract("", { "Chart.yaml": new ExtractorFile(content) });

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
