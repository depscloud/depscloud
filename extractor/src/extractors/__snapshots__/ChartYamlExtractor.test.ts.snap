// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartYamlExtractor fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "name": "beacon",
      "scopes": Array [],
      "versionConstraint": "https://mjpitz.github.io/charts/#0.0.2",
    },
    Object {
      "name": "mysql",
      "scopes": Array [
        "mysql.enabled",
      ],
      "versionConstraint": "https://charts.bitnami.com/bitnami#8.0.0",
    },
    Object {
      "name": "postgresql",
      "scopes": Array [
        "postgres.enabled",
      ],
      "versionConstraint": "https://charts.bitnami.com/bitnami#10.1.3",
    },
    Object {
      "name": "extractor",
      "scopes": Array [
        "extractor.enabled",
      ],
      "versionConstraint": "file://../extractor#0.2.33",
    },
    Object {
      "name": "gateway",
      "scopes": Array [
        "gateway.enabled",
      ],
      "versionConstraint": "file://../gateway#0.2.33",
    },
    Object {
      "name": "indexer",
      "scopes": Array [
        "indexer.enabled",
      ],
      "versionConstraint": "file://../indexer#0.2.33",
    },
    Object {
      "name": "tracker",
      "scopes": Array [
        "tracker.enabled",
      ],
      "versionConstraint": "file://../tracker#0.2.33",
    },
  ],
  "language": "helm",
  "name": "depscloud",
  "sourceUrl": "",
  "system": "helm",
  "version": "0.2.33",
}
`;

exports[`ChartYamlExtractor fullParse 2`] = `
"{
  \\"language\\": \\"helm\\",
  \\"system\\": \\"helm\\",
  \\"name\\": \\"depscloud\\",
  \\"version\\": \\"0.2.33\\",
  \\"sourceUrl\\": \\"\\",
  \\"dependencies\\": [
    {
      \\"name\\": \\"beacon\\",
      \\"versionConstraint\\": \\"https://mjpitz.github.io/charts/#0.0.2\\",
      \\"scopes\\": []
    },
    {
      \\"name\\": \\"mysql\\",
      \\"versionConstraint\\": \\"https://charts.bitnami.com/bitnami#8.0.0\\",
      \\"scopes\\": [
        \\"mysql.enabled\\"
      ]
    },
    {
      \\"name\\": \\"postgresql\\",
      \\"versionConstraint\\": \\"https://charts.bitnami.com/bitnami#10.1.3\\",
      \\"scopes\\": [
        \\"postgres.enabled\\"
      ]
    },
    {
      \\"name\\": \\"extractor\\",
      \\"versionConstraint\\": \\"file://../extractor#0.2.33\\",
      \\"scopes\\": [
        \\"extractor.enabled\\"
      ]
    },
    {
      \\"name\\": \\"gateway\\",
      \\"versionConstraint\\": \\"file://../gateway#0.2.33\\",
      \\"scopes\\": [
        \\"gateway.enabled\\"
      ]
    },
    {
      \\"name\\": \\"indexer\\",
      \\"versionConstraint\\": \\"file://../indexer#0.2.33\\",
      \\"scopes\\": [
        \\"indexer.enabled\\"
      ]
    },
    {
      \\"name\\": \\"tracker\\",
      \\"versionConstraint\\": \\"file://../tracker#0.2.33\\",
      \\"scopes\\": [
        \\"tracker.enabled\\"
      ]
    }
  ]
}"
`;
